### PR TITLE
Updates to Configurations for ICARUS MCP

### PR DESCRIPTION
**Pull request by gputnam**
Updates to ICARUS MC reconstruction for the upcoming MCP.

Right now includes a bugfix in running the stage1 MC reco fcl file and includes a number of configurations to run "CAF" making.

This is intended as a draft PR to collect updates needed for the MCP.

Depends on SBNSoftware/sbncode#156.

### ICARUS random number generator policy

To address the issue of `CORSIKAGen` (and `GENIEGen`?) requiring a random number when `NuRandomService` (`perEvent` policy) is not able to provide one, the policy for random number generation has been changed as follows.
* The general default policy in ICARUS is set to `random`.
* The policy for service configuration presets designed for simulation (`G4`, `DetSim` and also indirectly `Reco`) is set to `perEvent`, to preserve reproducibility.
* Selected hybrid job configurations that perform both generation and simulation tasks (e.g. in-time cosmic ray generation) have been explicitly set to `random` policy.

Note that integration test configurations for event generation already explicitly pin down the random number seeds, therefore they are not expected to be broken by this change.

As an aside, I have opened a [ticket with LArSoft](https://cdcvs.fnal.gov/redmine/issues/26116) to address the general issue with `perEvent` policies.

### MultiTPC DetSim configurations

They are the new `standard`.

### CODE REVIEW

**Comment by gputnam** [comment]
Also linked: https://github.com/SBNSoftware/icarusalg/pull/34

**Comment by PetrilloAtWork** [comment]
On Thursday, 5 August 2021 09:11:06 PDT Wesley Ketchum wrote:
> @wesketchum requested your review on: SBNSoftware/icaruscode#223 Updates to
> Configurations for ICARUS MCP.

Kind of conflict of interest here.
I should not be the only one reviewing, but I can concentrate on the parts I 
did not contribute to.

And let's see what happens when I try to answer to a GitHub notification e-
mail.

-- 
Gianluca Petrillo

SLAC building 084/B117, mail stop 78
Phone: +1 (650) 926-3088
E-mail: ***@***.***
Instant message: ***@***.*** (jabber)

**Comment by SFBayLaser** [comment]
I think this is critical for the next release, not the current one (v09_28_00) and so will leave open. I think there is still more work for the MC versions of stage0 and stage1. TBC.

**Comment by brucehoward-physics** [comment]
Pushing commit adding showers. Ran very small tests on a few particle gun MC events just to see that things ran basically. Possibly interestingly, in a test of a few events trying to put in SCE and SCE corr, I found very little difference in the sample having or not having the correction applied. It was not say 0 change for all showers, but in any case I wasn't sure if it should be as small a difference as it was. (just noted for future testing)

Also a few other things I believe I noticed from this test:
(1) This seemed to be running a different hit finding than I usually see. Was this intentional?
(2) I didn't scour the logs, mostly just was looking for it to have showers now. However, I think I noticed a few warnings/etc. from this test (I definitely noticed these, it's just a matter of it I noticed during _this_ test or another thing I was running, but I think it was this...)
* One was a warning that daqTrigger was invalid. Maybe related to using a particle gun generator fcl?
* "Failed to get GTruth object" - looks like while making a CAF file or one of the associated producer modules

**Comment by PetrilloAtWork** [comment]
I have pushed configurations to generate proton-only cosmic rays with overburden based on the "standard" configuration.
Unfortunately something seems broken in my box... cosmic rays generation fails when it's time to pick CORSIKA files.

**Comment by jzennamo** [comment]
I added a few missing generator level fcls needed for the August production and tested @PetrilloAtWork's proton-only cosmic configurations, things seem to be working!

**Comment by wesketchum** [comment]
Ooof, there are a lot of changes in here. Hard to track.

Can I recommend we try to cut off additions on this PR, move this from draft to non-draft, and make new PRs for any changes?? @gputnam @SFBayLaser ?

**Comment by gputnam** [comment]
Many of the changes are interrelated and making PR's on top of PR's can be its own mess. There are still a couple outstanding changes that need to be added (PMT + CRT reconstruction). I don't know if the inclusion of the CRT + PMT low-level (hit/flash) reconstruction will go in this PR or another. But once it is included I will also need to add in Iker's Flash matching to the high level reconstruction. So I don't think this PR is ready to move out of draft yet.

**Comment by gputnam** [comment]
I have merged in two more feature branches: Tracy's to include latest updates to stage0 / stage1 processing and Jacob's to run Iker's flash matching. Should be ready! So I will make this not a draft anymore.

I do want to run a couple more tests before this is merged, but it is ready for review.

**Comment by jzennamo** [comment]
This all looks awesome and I am really excited to have it merged in!

**Review by PetrilloAtWork — CHANGES_REQUESTED** [review]
My partial review (conflict of interest, remember?) found a couple of potential issues.

1. Naming of a few of new proton-only configurations does not follow the existing pattern.
2. Hard-coded numbers replicated around: _suggesting_ centralisation.
3. PMT decoding appears to have disappeared from one of the configurations.
4. one of the job configurations, `stage0_single_icarus.fcl`, seems to fail parsing.

Also there is material for a feature request to `GENIEGen` (it is possible to implement it without even changing `GENIEHelper`, which means that we can hope to see it soon).

File: fcl/caf/cafmakerjob_icarus.fcl [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  NuRandomService: @local::per_event_NuRandomService
+}
+
+#Start each new event with an empty event.
+source:
+{
+  module_type: RootInput
+  maxEvents:   -1        # Number of events to create
+}
+
```
[the remark this answers is not in the document]
Comment (gputnam) [reply fcl/caf/cafmakerjob_icarus.fcl:31]: Fixed.

File: fcl/caf/cafmakerjob_icarus.fcl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  outpid:
+  {
+    dataTier: cafana # for grid test
+    streamName: out1 # for grid test
+    module_type: RootOutput
+    fileName:    "pid_r%r_s%s.root"
+    fastCloning: false
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (gputnam) [reply fcl/caf/cafmakerjob_icarus.fcl:43]: Fixed the bad config -- although the block is used in one of the caf flavors so I am keeping it there.

**Comment by gputnam** [comment]
Added some commits to address Gianluca's comment and add in CRT hits/tracks to CAF-making

**Comment by gputnam** [comment]
Also I have run the tests that I wanted to!

File: fcl/gen/numi/simulation_genie_icarus_numi_volDetEnclosure_tpc.fcl [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+outputs.rootoutput.fileName: "simulation_genie_icarus_numi_volDetEnclosure_tpc_%tc-%p.root"
+
+# mbox is set as "xmin,ymin,zmin,xmax,ymax,zmax"
+# Add 10cm on side of each dimmension for some wiggle
+#
+# Because this is hardcoded, it needs to be updated for 
+# updates to the geometry. This is updates as of 
+# geometry: icarus_complete_20210527_no_overburden.gdml
+physics.producers.generator.FiducialCut: "mbox: -378.49,-191.86,-904.950652270838,378.49,144.96,904.950652270838"
```
[the remark this answers is not in the document]
Comment (gputnam) [reply fcl/gen/numi/simulation_genie_icarus_numi_volDetEnclosure_tpc.fcl:22]: We can include this as a future fix but it's still not clear to me that this is the best way to generate genie events in both cryostats, so I'd like to wait before optimizing the config.

**Comment by SFBayLaser** [comment]
trigger build

**Review by SFBayLaser — APPROVED** [review]
I am already familiar with the bulk of the changes and have verified they are working. There are further updates that I think I need to see in action. So I will merge this and then run through my checks on my laptop.

**Comment by SFBayLaser** [comment]
Ok, the expected outcome of the build since we are still getting things fixed there too. So will merge.

**Pull request merged**